src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -809,6 +859,42 @@ private void MoveDownCustomComparer((TElement Element, TPriority Priority) node,
             nodes[nodeIndex] = node;
         }
 
+        /// <summary>
+        /// Scans the heap for the first index containing an element equal to the specified parameter.
+        /// </summary>
+        private int FindIndex(TElement element, IEqualityComparer<TElement>? equalityComparer)
+        {
+            equalityComparer ??= EqualityComparer<TElement>.Default;
+            (TElement Element, TPriority Priority)[] nodes = _nodes;
+            int size = _size;
+
+            // Currently the JIT doesn't optimize direct EqualityComparer<T>.Default.Equals
+            // calls for reference types, so we want to cache the comparer instance instead.
+            // TODO https://github.com/dotnet/runtime/issues/10050: Update if this changes in the future.
+            if (typeof(TElement).IsValueType && equalityComparer == EqualityComparer<TElement>.Default)
+            {
+                for (int i = 0; i < size; i++)
+                {
+                    if (EqualityComparer<TElement>.Default.Equals(element, nodes[i].Element))
```

Comment:
How important is the performance of Remove?  If it's really important, we could consider subsequently changing the data storage of PQ, to be two arrays (priorities and elements) rather than one that combines them. Then here this could use IndexOf and get the SIMD benefits where applicable. (Did we try such splitting in the original prototyping/implementation of PQ? In theory it could have benefits even for normal enqueue/dequeue operations, as the priorities we'd be sifting through would be more tightly packed.)

Reply:
The original prototype did use separate arrays, IIRC we ended up fusing into one to optimize the regular heap operations and enumerator. I don't think the performance of `Remove` is important enough to override the original concerns.